lib/LaTeXML/Package/TeX.pool.ltxml

```diff
@@ -6758,28 +6756,32 @@ DefConstructor('\@XMRef{}', "<ltx:XMRef _xmkey='#1'/>",
 # but usually the XMRef is created first, and we want to let the referred to element
 # get it's id computed by whatever means it prefers.
 # so we have to work both ways (use state to record associations, to avoid expensive xpath)
+# Set id's on any non-XMRef nodes that have an _xmkey
+# This gets a more natural ordering
 Tag('ltx:*', 'afterOpen:late' => sub {
     my ($document, $node) = @_;
     if (my $key = $node->getAttribute('_xmkey')) {
       my $qname = $document->getNodeQName($node);
-      # If this is an XMRef w/o idref & referred node already has id; update the XMRef
-      if ($qname eq 'ltx:XMRef') {
-        if (!$node->getAttribute('idref')) {    # If not yet set
-          if (my $id = LookupValue('xmkey:' . $key)) {    # If XMArg already has an id
-            $document->setAttribute($node, idref => $id); }    # Note that id here
-          else {
-            PushValue('xmkey_referrers:' . $key, $node); } } }    # else, note that we still need an i
-          # If this is any other node w/_xmkey, make sure it has an id
-          # and record it on the corresponding XMRef
-      elsif ($qname =~ /^ltx:XM/) {
-        GenerateID($document, $node, undef, '');    # Generate id if none already.
-        my $id = $node->getAttribute('xml:id');
-        AssignValue('xmkey:' . $key => $id, 'global');    # for future use by XMRef
-              # but if we came after any XMRef's record their idref's as well.
-        if (my $referrers = LookupValue('xmkey_referrers:' . $key)) {
-          map { $document->setAttribute($_, idref => $id) } @$referrers;
-          AssignValue('xmkey_referrers:' . $key, undef, 'global'); } }
-} });
+      if (($qname ne 'ltx:XMRef') && ($qname =~ /^ltx:XM/) && !$node->hasAttribute('xml:id')) {
+        GenerateID($document, $node, undef, ''); } } });
+
+Tag('ltx:XMDual', 'afterClose:late' => sub {
+    my ($document, $node) = @_;
+    my %ids  = ();
```

Comment:
It is much nicer to see this lexical `%ids` compared to the previous State-based assignment/lookup. Seems like a win.

Reply:
I believe the What Was I Thinking rationale was thinking of XMRef as for more general sharing than only within XMDual. Only over time did it become clear that any non-dual usage needed different treatment (and probably a different element).